Dot_Marketplace-phase2 (#832)
w3f/Grants-Program: applications/dot_marketplace-phase2.md

| 0a | License | Apache 2.0 |
| 0b | Documentation | We will provide both inline documentation of the code and a tutorial that explains how a user can use DOT Marketplace and understand the flow of the tasking pallet. |
| 0c | Testing Guide | Functions will be covered by unit tests, the documentation will describe how to run these tests. We will also provide scripts to help deploy, run and test the build. |
| 0d | Docker Image | Docker image of the build |
| 1 | Migrate [Tasking Pallet](https://github.com/WowLabz/tasking_backend/blob/11ff1dfe620016d2943adc7b7a0ba60f2d6413cd/pallets/pallet-tasking/src/lib.rs) from FRAME v1 to FRAME v2 | The [existing Tasking Pallet](https://github.com/WowLabz/tasking_backend/blob/Phase1_Milestone3/pallets/pallet-tasking/src/lib.rs) will be migrated to FRAME v2 to account for the new features provided by the framework |
| 2 | Chat Pallet | Chat functionality is to be exposed and consumed between the customer and the service provider to ease communication and this will be integrated with the tasking pallet's frame v2 version |

### **Milestone 2**

* **Estimated duration:** 2 weeks
* **FTE:**  2
* **PTE:**  2
* **Costs:**  15,900 USD  

> In continuation to previous work, this milestone involves the creation of an on-chain decentralized court to handle dispute resolution. Being a juror is one of the user incentives that can be achieved thanks to the rating module mentioned in the first phase of dot marketplace. The entire milestone covers the Rust/Substrate code implementation. The court process can be called at any one of the instances mentioned from point 1a to 1c below. The dispute logic is a function which will be called from two extrinsics. The additional dispute function for 1c will to cover one case (at the moment- for ex, the customer is not closing the task due to some reason. The worker gets the functionality to raise dispute for this particular case). There can be more cases as such, which can be consumed and utilized for any use case. The functions for jury, that is court summon, dispute resolution time and other helpers can be configured and managed as per the use case.

Reviewer: ```suggestion
In continuation to previous work, this milestone involves the creation of an on-chain decentralized court to handle dispute resolution. Being a juror is one of the user incentives that can be achieved thanks to the rating module mentioned in the first phase of dot marketplace. The entire milestone covers the Rust/Substrate code implementation. The court process can be called at any one of the instances mentioned from points 1a to 1c below. The dispute logic is a function which will be called from two extrinsic. The additional dispute function for 1c will cover one case (at the moment- for example, the customer is not closing the task due to some reason. The worker gets the functionality to raise a dispute for this particular case). There can be more cases as such, which can be consumed and utilized for any use case. The functions for jury, which are court summon, dispute resolution time and other helpers, can be configured and managed for each use case.
```

- We are in touch with Marcin and Raul from the Web 3 Grants and Treasuries team, respectively.

## **Development Roadmap** 🔩

****Overview****

* **Total Estimated Duration:** 3 Weeks
* **Full-Time Equivalent (FTE):**  3.36
* **Total Costs:** 24,305 USD

### **Milestone 1**

* **Estimated duration:** 1 week
* **FTE:**  2
* **PTE:**  2
* **Costs:**  8,325 USD  

The main deiverable for this milestone will be to migrate the existing application to substrate frame v2 and include the chat feature as a pallet for communication between a customer and a service provider to have a one-on-one conversation over the deliverables and timelines. The entire milestone covers the Rust/Substrate code implementation.

| Sr no. | Deliverable | Description |
| --- | --- | --- |
| 0a | License | Apache 2.0 |
| 0b | Documentation | We will provide both inline documentation of the code and a tutorial that explains how a user can use DOT Marketplace and understand the flow of tasking pallet. |
| 0c | Testing Guide | Functions will be covered by unit tests, the documentation will describe how to run these tests. We will also provide scripts to help deploy, run and test the build. |

> | 0d | Docker Image | Docker image of the build |

Reviewer: Can you be a little bit more specific? What will the docker image contain?

Reviewer: @varun24143 could you add these details to the application as well, please?

| 2 | Chat Pallet | Chat functionality is to be exposed and consumed between the customer and the service provider to ease communication and this will be integrated with the tasking pallet's frame v2 version |
|  |  |  |

### **Milestone 2**

* **Estimated duration:** 2 weeks
* **FTE:**  2
* **PTE:**  2
* **Costs:**  15,900 USD  

In continuation to previous work, this milestone involves creation of a on chain decentralized court to handle dispute resolution. Being a juror is one of the user incentives that can be achieved thanks to the rating module mentioned in the first phase of dot marketplace. The entire milestone covers the Rust/Substrate code implementation.

| Sr no. | Deliverable | Description |
| --- | --- | --- |
| 0a | License | Apache 2.0 |
| 0b | Documentation | We will provide both inline documentation of the code and a tutorial that explains how a user can use DOT Marketplace and understand the flow of tasking pallet. |
| 0c | Testing Guide | Functions will be covered by unit tests, the documentation will describe how to run these tests. We will also provide scripts to help deploy, run and test the build. |

> | 0d | Docker Image | Docker image of the build |

Reviewer: Also here: What will the docker image contain?

Author: > Also here: What will the docker image contain?

The docker image will contain the entire tasking pallet Frame V2 version for anybody to just deploy it on their terminal without cloning the repo explicitly, it will be explained along with the commands for testing and running the image.

Reviewer: > to just deploy it on their terminal without cloning the repo explicitly

So by that you mean you'll upload it to a docker registry like docker hub?!

Feel free to add this info also to the grant application. In general, the application should contain all the details relevant to the application.

Author: @takahser Updated the application document. Please check and let me know.

> | 1 | Decentralized Court Module | An on chain decentralized court for dispute resolution within the ecosystem. |

Reviewer: ```suggestion
| 1 | Decentralized Court Module | An on-chain decentralized court for dispute resolution within the ecosystem. |
```

| 1a | Disapprove Task  | In the case of a customer not being satisfied by the work submitted by the service provider (worker). A set of jurors is formed (court-summon) to resolve the dispute and pass a verdict. |
| 1b | Disapprove Rating | The customer or the service provider, once they have received their rating for a particular task and are not satisfied by it. |

> | 1c | General Dispute | A general dispute function for cases that do not fall under the categories mentioned in 1a and 1b. |

Reviewer: I took a look at the Tasking Workflow Diagram where you show that the Court process can be started from 3 different starting points:

1. Approve Task by Customer: No
2. Worker satisfied by rating: No
3. Customer satisfied by rating: No

I would assume that the court process references these starting points wherever there is a dispute, is that correct? Hence, the dispute logic for 1a-1c would be the same, is that correct? I'm a bit confused why you'd implement an additional dispute function for 1c. Please feel free to correct me if I'm wrong.

Author: Yes, you are correct regarding the court process about being called at any one of these instances. The dispute logic is the same throughout the calls. The only difference is the scenarios when the dispute is called. The dispute logic is a function which is called from two extrinsics. The additional dispute function for 1c is to cover one case (at the moment), where lets say the customer is not closing the task due to some reason. The worker gets the functionality to raise dispute for this particular case. Since we thought there can be more cases as such, why not create an extrinsic which only raises a dispute.

Reviewer: Thanks for explaining, sounds good.
Maybe it would make sense to add this scenario to the application as well.

Reviewer: @varun24143 thanks for [adding this](https://github.com/w3f/Grants-Program/pull/832/files#diff-cbf2c9be64c2402648e5f355330c8ca6221cf239b8bfcaf95e67da105bd66062R179) to the doc. Feel free to add a reference/link to your comment in the future, this makes it easier for us to review what exactly got changed and might speed up the review process.

> | 2 | Jury | The chain specification of the testnet is modified to include more users with necessary specifications to be a part of the jury. The specifications include having average user rating above a certain threshold and being an expert in the field of the task. A list of potential jurors are notified and they have a period of one day to accept jury duty, with the maximum number of juors capped to 5 per dispute. |

Reviewer: > having average user rating above a certain threshold

1. Will this threshold be configurable or hardcoded? If configurable, how can it be changed?
2. How many jurors have to be elected for settling a dispute? Is it configurable? If configurable, how can it be changed?
3. What happens if there are not enough jurors that decide to become a juror within the 3d time period?
4. What happens if there are not enough jurors that decide to cast their vote?

Author: 1. The average threshold at the moment (i.e. 4) is hardcoded. We can provide a functionality for setting the criteria for average rating. And this can be set when raising the dispute.
2. Atleast 1 juror is required to settle the dispute. The max cap of jurors accepting jury duty is set, but it also can be added as a configuration when registering a dispute.
3. The program will do the calculations from among the jurors who participated
4. Same as 3

Reviewer: Ok thx. Also here, I think it'd be a good idea to integrate this information into the application document.

Author: @takahser Updated the application document. Please check and let me know.

Author: In the description of Milestone 2 - [general dispute and court schedule requirements](https://github.com/w3f/Grants-Program/blob/78d842b1d9fd11aaed20447eaa1015033a7b2056/applications/dot_marketplace-phase2.md#milestone-2)

> | 3 | Voting module | Each juror can review the dispute and cast their vote which also includes their rating for both the customer and the worker. After a period of two days all the juror votes are counted and a winner is identified. |

Reviewer: Here you say 2d, in the diagram it's 3d. Please clarify which one is the correct value and update accordingly.

Reviewer: Ideally, you'd even make it configurable.

Author: 3 days is the total time period for the entire court dispute. After the dispute is registered. 1 day is allotted for only accepting jury duty and after 1 day, the voting period starts which lasts for 2 days. And after 2 days the court is adjourned. And in those 2 days only the people who have accepted jury will be allowed to cast their votes.
The days can be configured based on how complex the task is. So we can provide a functionality before raising dispute, that for how many days does the raising party wants the trial to run.
So yes, it can be configured.
